modules/dhcp_common/subnet_service.rb

```diff
@@ -35,22 +41,44 @@ def add_subnets(*subnets)
     end
 
     def delete_subnet(subnet_address)
-      m.synchronize { subnets.delete(subnet_address) }
+      m.synchronize { subnets.delete(ipv4_to_i(subnet_address)); subnet_keys.delete(ipv4_to_i(subnet_address)) }
       logger.debug("Deleted a subnet: #{subnet_address}")
     end
 
     def find_subnet(address)
       m.synchronize do
-        to_ret = subnets[address]
-        return to_ret if to_ret # we were given a subnet address
+        ipv4_as_i = ipv4_to_i(address)
+        return subnets[ipv4_as_i] if subnets.key?(ipv4_as_i)
+        do_find_subnet(subnet_keys.to_a, subnets, ipv4_as_i, address)
+      end
+    end
 
-        # TODO: this can be done much faster
-        subnets.values.each do |subnet|
-          return subnet if subnet.include?(address)
-        end
+    def do_find_subnet(a_slice, all_subnets, address_as_i, address)
+      return nil if a_slice.empty?
+      if a_slice.size == 1
+        return all_subnets[a_slice.first].include?(address) ? all_subnets[a_slice.first] : nil
       end
+      if a_slice.size == 2
+        return all_subnets[a_slice.first] if all_subnets[a_slice.first].include?(address)
```

Comment:
Replacing the `Subnet#include?` calls to avoid IPAddr would probably be a good idea as you suggest. Creating new IPAddr objects from the `address` argument is taking approximately 25% of the wall time when calling `find_subnet` 1M times.